### ISSUE TITLE
kafka2 - Exception on write()

### ISSUE THREAD

**Issue by kevinkirkup** [fluent/fluent-plugin-kafka#263]
When attempting to write a chunk using the `kafka2` output plugin, I'm getting the following exception:

```
2019-05-28 19:53:46 +0000 [warn]: #0 Send exception occurred: undefined method `[]' for nil:NilClass
2019-05-28 19:53:46 +0000 [warn]: #0 Exception Backtrace : /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_kafka2.rb:206:in `write'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1125:in `try_flush'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1425:in `flush_thread_run'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:454:in `block (2 levels) in start'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-05-28 19:53:46 +0000 [info]: #0 initialized kafka producer: fluentd
2019-05-28 19:53:46 +0000 [warn]: #0 got unrecoverable error in primary and no secondary error_class=NoMethodError error="undefined method `[]' for nil:NilClass"
  2019-05-28 19:53:46 +0000 [warn]: #0 suppressed same stacktrace
2019-05-28 19:53:46 +0000 [warn]: #0 bad chunk is moved to /tmp/fluent/backup/worker0/object_2afda776ef10/589f80583e7651ee41a4734a66535ade.log
```

Looks like the code assumes that `chunk.metadata.variables` is always initialized, but the default is `nil`. Need to add handling for that.

**Comment by repeatedly** [comment]
Do you use v0.12 chunk files with kafka2?

**Comment by kevinkirkup** [comment]
No, I'm using v1.0.

Looks like this was fixed for the `rdkafka2`
https://github.com/fluent/fluent-plugin-kafka/blob/c6ec4331c7f16b2800804f120e8a7ed8409d61aa/lib/fluent/plugin/out_rdkafka2.rb#L233

**Comment by kevinkirkup** [comment]
After that, I'm seeing a different issue, however
```
2019-05-28 20:56:19 +0000 [warn]: #0 Send exception occurred: Failed to send messages to container_state/instance-0000007b, container_state/instance-0000007d, container_state/instance-0000007f, container_state/instance-00000081, container_state/instance-00000083, container_state/instance-00000086
2019-05-28 20:56:19 +0000 [warn]: #0 Exception Backtrace : /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/kafka_producer_ext.rb:244:in `deliver_messages_with_retries'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/kafka_producer_ext.rb:130:in `deliver_messages'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_kafka2.rb:230:in `write'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1125:in `try_flush'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1425:in `flush_thread_run'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:454:in `block (2 levels) in start'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-05-28 20:56:19 +0000 [info]: #0 initialized kafka producer: fluentd
2019-05-28 20:56:19 +0000 [debug]: #0 taking back chunk for errors. chunk="589f8e4edcc4f17fbe28debc39ad7c94"
2019-05-28 20:56:19 +0000 [warn]: #0 failed to flush the buffer. retry_time=1 next_retry_seconds=2019-05-28 20:56:20 +0000 chunk="589f8e4edcc4f17fbe28debc39ad7c94" error_class=Kafka::DeliveryFailed error="Failed to send messages to container_state/instance-0000007b, container_state/instance-0000007d, container_state/instance-0000007f, container_state/instance-00000081, container_state/instance-00000083, container_state/instance-00000086"
```

**Comment by kevinkirkup** [comment]
Looks `kafka` is expecting a partition number instead of a `partition_key` which is how the output plugin is being configured.

```
2019-05-28 20:56:19 +0000 [warn]: #0 Send exception occurred: Failed to send messages to container_state/instance-0000007b, container_state/instance-0000007d, container_state/instance-0000007f, container_state/instance-00000081, container_state/instance-00000083, container_state/instance-00000086
2019-05-28 20:56:19 +0000 [warn]: #0 Exception Backtrace : /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/kafka_producer_ext.rb:244:in `deliver_messages_with_retries'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/kafka_producer_ext.rb:130:in `deliver_messages'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_kafka2.rb:230:in `write'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1125:in `try_flush'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1425:in `flush_thread_run'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:454:in `block (2 levels) in start'
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-05-28 20:56:19 +0000 [info]: #0 initialized kafka producer: fluentd
2019-05-28 20:56:19 +0000 [debug]: #0 taking back chunk for errors. chunk="589f8e4edcc4f17fbe28debc39ad7c94"
2019-05-28 20:56:19 +0000 [warn]: #0 failed to flush the buffer. retry_time=1 next_retry_seconds=2019-05-28 20:56:20 +0000 chunk="589f8e4edcc4f17fbe28debc39ad7c94" error_class=Kafka::DeliveryFailed error="Failed to send messages to container_state/instance-0000007b, container_state/instance-0000007d, container_state/instance-0000007f, container_state/instance-00000081, container_state/instance-00000083, container_state/instance-00000086"
```

**Comment by kevinkirkup** [comment]
Use 
```
root@virt-pod-17a-e2:/home/vagrant# td-agent --version
td-agent 1.4.2
```

APT Info
```
root@virt-pod-17a-e2:/home/vagrant# apt-cache policy td-agent
td-agent:
  Installed: 3.4.1-0
  Candidate: 3.4.1-0
  Version table:
 *** 3.4.1-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
        100 /var/lib/dpkg/status
     3.4.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.3.0-1 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.3.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.2.1-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.2.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     2.5.0-0 500
        500 http://packages.treasuredata.com/2.5/ubuntu/bionic bionic/contrib amd64 Packages
```

**Comment by kenhys** [comment]
Fixed via #265 and released as v0.9.4.
 

```
Release 0.9.4 - 2019/05/13

	* out_rdkafka2: Add metadata nil check
```
